### ISSUE TITLE
Proposal: intelligently suggest pythonPath where possible for conda/virtualenv users

### ISSUE THREAD

**Issue by jabooth** [DonJayamanne/pythonVSCode#257]
I was thinking of making it a little easier to get started with the extension for the case of users with multiple Python environments.

@DonJayamanne I saw you were asking for some feedback on how to set the `"python.pythonPath"` for the case of having multiple python environments in https://github.com/DonJayamanne/pythonVSCode/issues/61#issuecomment-226636855. With that being a closed issue, I thought I would offer my suggestion here, and if you like the direction (and I find the time) I could maybe drop a PR for this.

Here's my use case - we use [miniconda](http://conda.pydata.org/miniconda.html) envs, and I normally have a dozen or so on my machine, all for different projects. I like the elegance of just launching code from the command line with the right interpreter activated, but in practice I think it's never going to be a common paradigm for users of this plugin for a number of reasons:
1.  It's very natural to launch Code with the OS GUI launcher (OS X Spotlight, Windows Start, Ubuntu Dashboard)
2. The default install does not add `code` to the PATH
3. It's natural to open new windows from within code (shift + cmd + N) to open a new project

Because of this, I think it's more likely that users of this extension will need to set the `pythonPath` from within Code itself on a per-project basis. This is does have a number of benefits:
1. The choice is persisted beyond code restarts (set it once on a machine and it can stay for a while)
2. `.vscode/settings.json` can be added to `.gitignore` so each developer has their own per-project python interpreter saved without collision (`/Users/jab08/.../bin/python` vs `/home/patrick/.../bin/python`)
3. You can (in theory [1]) change the Python interpreter without restarting code, e.g. to check Python 2.7/3.5 compatibility.

My proposal would be to add a new command to the extension (accessible within the Command Palette) called something like: `Python: Set Workspace Interpreter`. Hitting enter on this, you would be presented with a shortlist of suggested Python interpreters found on this system. This would be filterable with typing in a few letters (similar to the old Install Extensions interface) and once a selection is made, hitting enter would save the chosen interpreter to `.vscode/settings.json`. It would be nice if at the top it showed the currently set interpreter too.

We could detect common patterns in the users path that would suggest they are using a conda or virtualenv setup, and then based on this parse the relevant dirs (e.g. `~/miniconda/envs/*/bin/python`) to generate the listings. This should be pretty fast, and of course we would do it dynamically when the user launches the command. We could also look at PyCharm for some inspiration, as they also [suggest interpreters for the user](https://www.jetbrains.com/help/pycharm/2016.1/configuring-available-python-interpreters.html).

I haven't worked on a vscode extension yet, but if you like the idea I'm happy to give it a go and stick up a PR?

[1] I note [here](https://github.com/DonJayamanne/pythonVSCode/wiki/Python-Path-and-Version#virtual-environments) you mention hot changing of interpreters is something that will be supported in a future version

**Comment by DonJayamanne** [comment]
Brilliant suggestion, I like that, besides it just adds functionality without altering any thing existing. Please proceed with a PR

**Comment by jabooth** [comment]
Hmm surprisingly there is no extension API for writing to the workspace settings for now: https://github.com/Microsoft/vscode/issues/1396. However, it is under active discussion and looks like it could land in the next few months.

I will still prepare the PR now with a shim around the actual writing of the settings. Then we can wait for the official API to drop before merging, or we could see how robustly we can manually adjust the settings ourselves as a stop-gap (my first thought is probably better to wait!)

I suppose another nice alternative (which we could add regardless) is we could offer the same found interpreters as autocomplete suggestions upon editing the `settings.json` file manually, in the same vein as:
<img width="716" alt="screenshot 2016-08-21 10 31 11" src="https://cloud.githubusercontent.com/assets/1312873/17836425/783483be-678a-11e6-842e-21a7d8fb12da.png">
saying that, I don't know where these suggestions come from and if they are dynamically extensible? My first guess is they are just typescript suggestions based on the type info for each configuration field, and hence it's probably not possible to hook in here?

**Comment by DonJayamanne** [comment]
This adds a new Command Palette entry for Python which provides suggestions for Python interpreters you may wish to set in the workspace settings:
![Python: Set Workspace Interpreter](https://cloud.githubusercontent.com/assets/1312873/17837213/8060763c-67a4-11e6-9a06-cc71c23e4a8f.png)


For now, only conda environments are detected (by interrogating the conda binary), but it will be easy to add support for other common Python interpreter locations.

See #257 for more motivation.

VSCode doesn't yet include support for setting workspace settings from an extension, so for now just a message is printed to demonstrate where we would update `.vscode/settings.json`.

#### Todo

As a minimum before this can be merged:
- [x] Test on Windows
- [x] Test in absence of conda
- [ ] Add documentation ([here on the wiki](https://github.com/DonJayamanne/pythonVSCode/wiki/Python-Path-and-Version#virtual-environments) I guess?)
- [ ] Add default Python install locations (e.g. `/usr/local`)

#### Extensions

Could be handled in separate PRs:
- [x] Test for existence of suggestions and filter based on existence
- [ ] Detect Python version and include in hint
- [ ] Add virtualenv detection
- [x] Add ability to copy selected path to clipboard
- [ ] Actually set the workspace setting (blocked by Microsoft/vscode#1396, could write our own as stopgap?)
- [ ] Add unit tests for new functionality

**Comment by DonJayamanne** [comment]
Closing this issue, will update the documentation later today/tomorrow before the release.

**Comment by StevenCHowell** [comment]
> - [ ]  Add documentation (here on the wiki I guess?)

I tried searing in the Command Palette for my conda environments but it does not find anything.  I checked the [wiki](https://github.com/DonJayamanne/pythonVSCode/wiki/Python-Path-and-Version) but do not see any pertinent information.  From the unchecked box, it seems it has not yet been added.  How do I use this feature?  Do I need to install anything other than @DonJayamanne's Python extension?

(p.s. I enjoyed the [VS Code TalkPython podcast episode](https://talkpython.fm/episodes/show/101/adding-a-full-featured-python-environment-to-visual-studio-code))
